core/src/main/java/nl/nn/adapterframework/webcontrol/api/TransactionalStorage.java

```diff
@@ -85,21 +93,29 @@ public Response browseReceiverMessage(
 			throw new ApiException("Adapter not found!");
 		}
 
-		Receiver<?> receiver = adapter.getReceiverByName(receiverName);
-		if(receiver == null) {
-			throw new ApiException("Receiver ["+receiverName+"] not found!");
+		String message = null;
+		IMessageBrowser<?> storage = null;
+		messageId = Misc.urlDecode(messageId);
+
+		if(STORAGESOURCE_TYPE_PIPE.equals(storageSource)) {
+			MessageSendingPipe pipe = (MessageSendingPipe) adapter.getPipeLine().getPipe(storageSourceName);
+			if(pipe == null) {
+				throw new ApiException("Pipe ["+storageSourceName+"] not found!");
+			}
+			storage = pipe.getMessageLog();
+			message = getMessage(storage, messageId);
+		} else {
+			Receiver<?> receiver = adapter.getReceiverByName(storageSourceName);
+			if(receiver == null) {
+				throw new ApiException("Receiver ["+storageSourceName+"] not found!");
+			}
+			storage = receiver.getMessageBrowser(ProcessState.getProcessStateFromName(processState));
+			message = getMessage(storage, receiver.getListener(), messageId);
```

Comment:
Thanks for combining this code!
This code, that is repeated similarly below, could benefit from defining an interface IMessageStorageOwner or so.
Then it would/should be enough to call
```
IMessageStorageOwner storageOwner = getMessageStorageOwner(storageSource, storageSourceName);
storageOwner.<doTheThingsRequired>;
```